### ISSUE TITLE
Compatibility: The plugin should try old constructors if the existing DataBound one does not match the fields

### ISSUE THREAD

**Issue by oleg-nenashev** [jenkinsci/configuration-as-code-plugin#233]
Currently the CasC plugin does not offer compatibility for old-style logic using `@DataBoundConstructor`. If a plugin configuration gets updated (new constructor, old is deprecated), the plugin just fails. It leads to errors like #167 (change of work dir settings, CC @MadsNielsen ), which appear in valid cases from UI perspective. 

I propose to extend the compatibility layer in DataBoundConfigurator to try old constructors OR to use default for newly introduced fields

**Comment by ndeloof** [comment]
We can't use non-@DataBound constructors as we have no way using java reflexion to guess argument names.
Also, let's consider why a new DataBound constructor is introduced by some refactoring:
1. new required parameter added. So you need to provide it, no other way
2. legacy option parameter moved to a `@DataBoundSetter`. Configuration then is still valid.

considering #167, this has been badly refactored imho. Making `workDirSettings` as non-null suggest this one can't have a default value, but legacy constructor still can find one. So would better just have been a DataBoundSetter.

**Comment by oleg-nenashev** [comment]
Badly refactored? Maybe. But plugin developers do many crappy things.

Generally I agree that supporting all constructors is a bad way. What if we introduce a `@DeprecatedDataBoundConstructor` annotation to indicate that a constructor used to be a DataBound one?

**Comment by ndeloof** [comment]
Configuration as Code is based on correct implementation of databinding patterns. Most plugin's Descriptor don't rely on `DataBoundSetter`s, but will do if they want to be supported. 
I'd like CasC to help promote good design practices, even this require some time for all plugins to be well designed accordingly. Adding flexibility to support all crapy things would just let developer use them for yet another decade. And as a short terms fix, they just can provide a custom `Configurator` implementation.

**Comment by oleg-nenashev** [comment]
So effectively it means that the plugin will never support `@DataBoundConstructor` upgrades in a compatible way, right? It would limit the plugin's usability a lot for plugins targeting old baselines, and even new plugins may break at any moment.

CC @ewelinawilkosz @MadsNielsen What would be your opinion?

**Comment by MadsNielsen** [comment]
Difficult question, i like the idea by @oleg-nenashev about introducing an `@DeprecatedDataBoundConstructor` to help developers clearly indicate that a new required parameter is being introduced. My guess in those cases, is that the developer already was following good practice so this will make documenting upgrades for them a lot easier as it can be done in a non breaking way by putting in a note with a new configuration sample. It also allows us to cleanly select an alternate constructor, ruling out those that were not meant for UI configuration in the first place.   

I can also see the point raised by @ndeloof that we can promote good design practice from developers with the use of Databund setters to cleanly deprecate fields. We want to avoid decade long, temporary fixes. 

It all comes down to ease of use and value for customers vs. promoting clean code in plugins, which is a tough balancing act. In order for it to be successful it either needs to have very clean auto generated documentation, with very clean example configurations so it's easy to figure out the new configuration in case something changes, or we have to code defensively to support as many use cases as possible without resorting to glue code. If we do use defensive coding, we have to support a lot more configurations which can become a problem over time if a user decides to use an ancient version of plugin X. 
 
Also while typing this it appears that @oleg-nenashev has come up with a commit that explores the use of non databound contructors...you guyes are fast! :+1:. Gonna check that out!

**Comment by oleg-nenashev** [comment]
@ndeloof Nope, I got it exactly as you mean. 

>  If you define a new DataBoundConstructor because you introduce a fresh new required attribute (which you can't define a reasonable default for) then there's nothing we can do.

I do disagree with that statement. Extending constructors is a perfectly valid plugin development approach which has never been officially deprecated. Yes we have `@DataBoundSetter` support now, but it does not make the constructor use-case obsolete. Moreover, `@DataBoundSetter` is not a thing which all plugin maintainers like, because it makes them to convert `final` fields to non-final and to introduce new Setter methods in API. If one likes the defensive code for configurations, it is not something they would like.

My proposal is to find a way to support `@DataBoundConstructor` updates, at least for some common cases. If you do not like `@DeprecatedDataBoundConstructor` or solution in #234, we can consider other options like field annotations (`@Optional` or `@CheckForNull`, for example).

Otherwise we may need to update Jenkins-wide development guidelines. It is possible, of course. But it would be rather another JEP to deprecate DataBound constructors.

**Comment by ndeloof** [comment]
introducing `@DeprecatedDataBoundConstructor` would require recent stapler and core, so won't be available for plugin developer before months (or years?).

I indeed agree my preference for @DataBoundSetters might not be shared with others, so we need to consider constructor upgrades for new optional attributes ... and the good news is that we already do : we support `@NonNull` to detect *required* constructor arguments, considering them optional by default (because way too much constructors in jenkins-core are designed this way).

So IMO https://github.com/jenkinsci/configuration-as-code-plugin/pull/168 was just wrongly annotated @Nonnull as this attribute has a reasonable default value.

**Comment by oleg-nenashev** [comment]
Yes. This part in the core is something we definitely need to change to unblock this use-case at least for JEP-200 testing

**Comment by oleg-nenashev** [comment]
https://issues.jenkins-ci.org/browse/JENKINS-51603 is WiP
